7 implement intake

j0shuaS opened the pull request:
this is a draft just to make sure that what I have so far is good. If @aidnem could take a look that would be amazing.

j0shuaS committed:
foundation of intake arm

j0shuaS committed:
foundation of intake arm

j0shuaS committed:
everything works exepct json.

j0shuaS committed:
everything works exepct json.

j0shuaS committed:
All functionality is there. Ready to update constants and put on a real robot!

j0shuaS committed:
All functionality is there. Ready to update constants and put on a real robot!

j0shuaS committed:
Added methods to control the movement of Intake subsystem

j0shuaS committed:
Added methods to control the movement of Intake subsystem

aidnem commented on src/main/java/frc/robot/subsystems/intake/IntakeRollerIOSim.java:
```diff
@@ -0,0 +1,85 @@
+public class IntakeRollerIOSim extends IntakeRollerIOTalonFX {
+  private TalonFXSimState motorSimState;
+
+  private FlywheelSim leftSim =
```
Nitpick: I would name this something other than left sim since there's only one

aidnem commented on src/main/java/frc/robot/subsystems/intake/IntakeRollerIOSim.java:
```diff
@@ -0,0 +1,85 @@
+    if (JsonConstants.intakeConstants.rollerInverted == InvertedValue.Clockwise_Positive) {
+      velocityCache.mut_times(-1.0);
+      accelCache.mut_times(-1.0);
+    }
```
@j0shuaS looks like you took my flywheel code in a broken state (or at least it seemed broken to me). If sim works for you, that's great, I'll copy some of this stuff. If sim doesn't work for you, you can look at my 8-implement shooter branch for how I fixed some of the weird issues.

aidnem commented on src/main/java/frc/robot/subsystems/intake/IntakeRollerMechanism.java:
```diff
@@ -0,0 +1,176 @@
+  private RollerSpeeds goalSpeeds = ZERO_SPEEDS;
+
+  private enum RollerOutputMode {
+    ClosedLoop,
```
Probably don't need a ClosedLoop or Current Mode since it's just rollers. You can probably just run it at a voltage and it should "just work"

aidnem commented on src/main/java/frc/robot/subsystems/intake/IntakeSubsystem.java:
```diff
@@ -0,0 +1,48 @@
+  private IntakeArmMechanism arm;
+  private IntakeRollerMechanism roller;
+
+  public IntakeSubsystem(IntakeArmMechanism arm, IntakeRollerMechanism roller) {
```
If you're using the "create" method architecture, the constructor should be private. That way, the only way to make one is through "create" and you can be sure that the protection for multiple instances is guaranteed to run.

aidnem commented on src/main/java/frc/robot/subsystems/intake/IntakeSubsystem.java:
```diff
@@ -0,0 +1,48 @@
+import coppercore.wpilib_interface.MonitoredSubsystem;
+import java.util.Optional;
+
+public class IntakeSubsystem extends MonitoredSubsystem {
```
You will probably need some sort of state machine or something for homing. I don't know if there's a CANcoder, so you might have to home the intake against the side of the shooter by driving up to the top until it stops moving.

aidnem commented on src/main/java/frc/robot/TestModeManager.java:
```diff
@@ -0,0 +1,71 @@
+    ShooterClosedLoopTuning("Shooter Closed Loop Tuning"),
+    IntakeRollerVoltageTuning("Intake Roller Voltage Open Loop Tuning"),
+    IntakeRollerCurrentTuning("Intake Roller Current Open Loop Tuning"),
+    IntakeRollerClosedLoopTuning("Intake Roller Closed Loop Tuning");
```
TODO: Add tuning modes for the intake arm too (you probably already have this planned but noting it anyway)

aidnem reviewed (commented):
Overall looks good to me. Tried to give a few pointers.

j0shuaS committed:
sim is broken

j0shuaS committed:
sim is broken

j0shuaS commented:
@PChild hii there is something really weird happening with sim on arm. I think it has something todo with units? Can you looks please

aidnem committed:
Implement Shooter (#15)

* (#8) Move shooter to its own folder and add ShooterIO

* (#8) WIP - Start adding shooter logic

* (#8) Use JsonConstants for ShooterConstants loading and implement some Shooter stubs

* (#8) Add basic docs on scoring subsystem and shooter mechanism

* (#8) Try to fix links in scoring docs

* (#8) Add ShooterIOTalonFX

* (#8) WIP (untested) - Add ShooterIOSim

* (#8) WIP - Sim not functional - Add Test mode

* (#8) Enable continuous closed-loop and apply gearing calculation, sim still doesn't work

* (#8) Potentially fix bug from reusing the same cached unit object in ShooterIOSim

* (#8) Make gradlew executable

* (#8) Try negating motor inputs in Sim IO, fixes DiscordantMotoring bug but now current draw is ridiculously high

* (#8) Refactor shooterIO to use 1 for each side, removing a ton of duplicated code

* (#8) Add comment about logging

* (#8) WIP - Simplify sim code. It now works but only CCW with gearing=1.0

* (#8) WIP - Shooter Sim now works with both motor inverts, gearing is still untested

* (#8) Add max acceleration tuning to shooter, tune shooter. Shooter sim now functions with closed-loop

* (#8) Add closed loop reference logging to ShooterIOTalonFX

* (#8) Make ShooterSpeeds into a StructSerializable class so it can be logged by akit

* (#8) Add pose-based shots to shooter

* (#8) Update shooter docs to reflect changed API and sync all JSON constants

* (#8) Format JSON

* (#8) Clarify shooter constant names (review requested)

* (#8) Make shooter distance->RPM maps automatically determine minShotDistance based on shortest distance in the map

* (#8) Fix review comments: clarify names, doc comments, method behavior.

- Clarify units in names of Shooter Constants
- Fix typo in ShooterMechanism.java
- Add javadocs to FieldLocations constant files
- Clarify behavior in GeomUtil.cross doc comment
- Improve shooter constants initialization method name

* (#8) Make tryUntilOk return whether or not it succeeded and make ShooterIOTalonFX print an error & display an alert when failing to apply a config

* (#8) Update shooter adoc to reflect that goal speeds are in terms of the motor

* (#8) Clarify ShooterIO.runSpeed doc comment to indicate that it is in terms of motor speed

* (#8) Revert deletion of PathPlanner folder because navgrid and settings may be used later for OTF pathing

* (#8) Make shooter only able to shoot left, add InitBindings file

* (#8) Add missing file, mistakenly didn't add it in last commit

* (#8) Rename rayDirection and shotDirection to rayDirectionVector and shotDirectionAngle for greater clarity

* (#8) Swap shooter epsilon to a constant value rather than a fraction

* (#8) Make PhoenixUtil status code name more descriptive; remove stray semicolon; format

* (#8) Improve docs to reflect changes & decisions made during development

* (#8) Add ScoringFeatureFlags and make shooter mechanisms Optional

* (#8) Remove TODO to figure out 0.0 motor voltage; I figured it out; it was the second ShooterIO printing its idle voltage

* (#8) Fix scoring.adoc typo: Interplating->Interpolating

* (#8) Make InitBindings final and give it private constructor

* (#8) Move CAN bus name to CANConstants

* (#8) Fix 60 RPS shooter velocity epsilon -> 1 rps (60 rpm)

* (#8) Fix bug where test modes were ignored due to shooter action enum and document shooter testing in sim

* (#8) Rename optional[Mechanism] to [mechanism] in shooter subsystem

---------

Co-authored-by: ahmed alkushi <231575845+401User2025@users.noreply.github.com>

j0shuaS committed:
changing can ids

j0shuaS committed:
changing can ids

j0shuaS committed:
Merge branch '7-implement-intake' of https://github.com/team401/Serpens-2025 into 7-implement-intake

j0shuaS committed:
rebased main

j0shuaS committed:
debugging

j0shuaS committed:
trying real constants

aidnem commented:
> @PChild hii there is something really weird happening with sim on arm. I think it has something todo with units? Can you looks please

@j0shuaS Did you ever figure this out? It looked like it was working on Wednesday

avidraccoon commented:
> @j0shuaS Did you ever figure this out? It looked like it was working on Wednesday

I believe it worked on Wednesday. And I think the problem was that he used the encoder to control the motor, but did not update the encoder's simulation state.

j0shuaS commented:
Yep it worked with those constants, working on putting real ones into play.
